confidential-data-hub/kms/src/plugins/aliyun/client/mod.rs

```diff
@@ -0,0 +1,274 @@
+
+#[derive(Clone, Debug)]
+enum Client {
+    ClientKey {
+        client_key_client: ClientKeyClient,
+    },
+    EcsRamRole {
+        ecs_ram_role_client: EcsRamRoleClient,
+    },
+}
+
+pub struct AliyunKmsClient {
+    inner_client: Client,
+}
+
+const ALIYUN_IN_GUEST_DEFAULT_KEY_PATH: &str = concatcp!(_IN_GUEST_DEFAULT_KEY_PATH, "/aliyun");
+
+impl AliyunKmsClient {
+    pub fn new(
+        client_key: &str,
+        kms_instance_id: &str,
+        password: &str,
+        cert_pem: &str,
+    ) -> Result<Self> {
+        Self::new_client_key_client(client_key, kms_instance_id, password, cert_pem)
+    }
+
+    pub fn new_client_key_client(
+        client_key: &str,
+        kms_instance_id: &str,
+        password: &str,
+        cert_pem: &str,
+    ) -> Result<Self> {
+        let inner_client = Client::ClientKey {
+            client_key_client: ClientKeyClient::new(
+                client_key,
+                kms_instance_id,
+                password,
+                cert_pem,
+            )?,
+        };
+
+        Ok(Self { inner_client })
+    }
+
+    pub fn new_ecs_ram_role_client(ecs_ram_role_name: &str, region_id: &str) -> Result<Self> {
+        let inner_client = Client::EcsRamRole {
+            ecs_ram_role_client: EcsRamRoleClient::new(ecs_ram_role_name, region_id)?,
+        };
+
+        Ok(Self { inner_client })
+    }
+
+    /// This new function is used by a in-pod client. The side-effect is to read the
+    /// [`ALIYUN_IN_GUEST_DEFAULT_KEY_PATH`] which is the by default path where the credential
+    /// to access kms is saved.
+    pub async fn from_provider_settings(provider_settings: &ProviderSettings) -> Result<Self> {
+        let inner_client = if provider_settings.contains_key("client_type")
+            && provider_settings["client_type"] == "client_key"
+        {
+            Client::ClientKey {
+                client_key_client: ClientKeyClient::from_provider_settings(provider_settings)
+                    .await
+                    .map_err(|e| {
+                        Error::AliyunKmsError(format!(
+                            "build ClientKeyClient with `from_provider_settings()` failed: {e}"
+                        ))
+                    })?,
+            }
+        } else if provider_settings.contains_key("client_type")
+            && provider_settings["client_type"] == "ecs_ram_role"
+        {
+            Client::EcsRamRole {
+                ecs_ram_role_client: EcsRamRoleClient::from_provider_settings(provider_settings)
+                    .await
+                    .map_err(|e| {
+                        Error::AliyunKmsError(format!(
+                            "build EcsRamRoleClient with `from_provider_settings()` failed: {e}"
+                        ))
+                    })?,
+            }
+        } else {
+            return Err(Error::AliyunKmsError(
+                "client type invalid or not exist.".to_string(),
+            ));
+        };
```

Comment:
Could you do some refactoring like the following to make the logic clearer
```suggestion
        let Some(client_type) = provider_settings.get("client_type") else {
                     return Err(Error::AliyunKmsError(
                "client type invalid or not exist.".to_string(),
            ));
        };
        
        let inner_client = match client_type {
            "client_key" => ...
            "ecs_ram_role" => ...
        }
```

Reply:
Great! It does look clearer.